File: musicApp/Sync/SettingsSync.swift
import Foundation
import Combine
import FirebaseFirestore

/// Two-way effects-settings sync — users/{uid}/sync/settings (singleton
/// doc, same doc family as the session/playlist docs). Twin of
/// desktop/src/settingsSync.ts — FIELD NAMES ARE THE CONTRACT.
///
/// LWW is Firestore's own snapshot ordering (same as SessionState/
/// CloudPlaylist) — `at` (ServerClock ms) is carried for parity/debugging,
/// not compared client-side. `updatedBy` filters same-device echo.
///
/// Speed and bass share units with desktop as-is (multiplier, dB). Reverb is
/// wired 0-100% on the wire; desktop's internal `fx.reverb` is a 0-1
/// fraction, converted at its own sync boundary.
///
/// `bypass` (iOS effectsBypass / desktop fx.bypass) is part of the doc — it is
/// OPTIONAL on the wire, and absent means "written by a pre-M10 client", read
/// as false. Without it a bypassed device published slider values it wasn't
/// hearing and a non-bypassed device applied them audibly, contradicting
/// PlaybackState.rate, which IS bypass-adjusted (sync-audit-4 M10).
///
/// iOS persists these per-track (`AudioPlayerManager.TrackSettings`), so
/// switching tracks changes the published values as a side effect — that's
/// intentional here: it syncs "whichever effective settings are currently
/// audible", the same way playback state syncs the currently-playing track.
@MainActor
final class SettingsSync {

    private let db: Firestore
    private let player: AudioPlayerManager
    private var bag = Set<AnyCancellable>()
    private var listener: ListenerRegistration?
    private var uid = ""

    // Last values WE applied from a remote snapshot. Lets the local publish
    // sink tell "user moved a slider" apart from "assigning the remote value
    // re-fired @Published" without a timing flag — the publish side is
    // debounced, so a flag would have cleared by the time it fires.
    private var lastAppliedSpeed: Double?
    private var lastAppliedBass: Double?
    private var lastAppliedReverb: Double?
    private var lastAppliedPitch: Double?
    private var lastAppliedBypass: Bool?

    /// True when THIS device owns the session's audio. Decides whether a
    /// remote's values are persisted into the current track's memory (see
    /// applyRemote). Injected so this class stays independent of the
    /// coordinator; SyncSessionManager wires it.
    private let ownsAudio: () -> Bool

    init(db: Firestore, player: AudioPlayerManager, ownsAudio: @escaping () -> Bool) {
        self.db = db
        self.player = player
        self.ownsAudio = ownsAudio

        // effectsBypass rides along (sync-audit-4 M10): publishing the sliders
        // without the master switch meant a bypassed device pushed values it
        // wasn't hearing and the other device played them — and they
        // contradicted PlaybackState.rate, which IS bypass-adjusted.
        //
        // Pitch rides along too. It was the one effect neither side published
        // ("reverb, pitch and bass don't sync") — nested CombineLatest because
        // CombineLatest4 is the widest Combine ships.
        Publishers.CombineLatest(
            Publishers.CombineLatest4(player.$playbackSpeed, player.$bassBoost,
                                      player.$reverbAmount, player.$effectsBypass),
            player.$pitchShift)
            .debounce(for: .milliseconds(300), scheduler: DispatchQueue.main)
            .sink { [weak self] core, pitch in
                let (speed, bass, reverb, bypass) = core
                self?.push(speed: speed, bass: bass, reverb: reverb, pitch: pitch, bypass: bypass)
            }
            .store(in: &bag)
    }

    func activate(uid: String) {
        self.uid = uid
        listener?.remove()
        listener = docRef.addSnapshotListener { [weak self] snap, _ in
            guard let snap else { return }
            Task { @MainActor in self?.applyRemote(snap) }
        }
    }

    /// Detach from the current home — otherwise `forgetHome` leaves this
    /// listener applying a forgotten home's effects (sync-audit-4 M9).
    /// lastApplied* is cleared so reconnecting re-pushes local values instead
    /// of believing the new home already has them.
    func deactivate() {
        listener?.remove(); listener = nil
        uid = ""
        lastAppliedSpeed = nil
        lastAppliedBass = nil
        lastAppliedReverb = nil
        lastAppliedPitch = nil
        lastAppliedBypass = nil
    }

    private var docRef: DocumentReference {
        db.collection("users").document(uid).collection("sync").document("settings")
    }

    private func push(speed: Double, bass: Double, reverb: Double, pitch: Double, bypass: Bool) {
        guard !uid.isEmpty else { return }
        if speed == lastAppliedSpeed, bass == lastAppliedBass,
           reverb == lastAppliedReverb, pitch == lastAppliedPitch,
           bypass == lastAppliedBypass { return }

        // Update lastApplied* to track "last state we believe Firestore already has"
        lastAppliedSpeed = speed
        lastAppliedBass = bass
        lastAppliedReverb = reverb
        lastAppliedPitch = pitch
        lastAppliedBypass = bypass

        let doc: [String: Any] = [
            "speed": speed, "bassDb": bass, "reverbPct": reverb, "pitchSt": pitch,
            "bypass": bypass,
            "updatedBy": SyncDevice.id, "at": ServerClock.shared.nowMs,
        ]
        Task { try? await docRef.setData(doc) }
    }

    private func applyRemote(_ snap: DocumentSnapshot) {
        guard let d = snap.data(),
              let by = d["updatedBy"] as? String, by != SyncDevice.id,
              let speed = (d["speed"] as? NSNumber)?.doubleValue,
              let bass = (d["bassDb"] as? NSNumber)?.doubleValue,
              let reverb = (d["reverbPct"] as? NSNumber)?.doubleValue else { return }

        let clampedSpeed = min(max(speed, 0.5), 2.0)
        let clampedBass = min(max(bass, -10), 20)
        let clampedReverb = min(max(reverb, 0), 100)
        // Absent = written by a client from before pitch synced; read as 0.
        let pitch = (d["pitchSt"] as? NSNumber)?.doubleValue ?? 0
        let clampedPitch = min(max(pitch, -12), 12)
        // Absent = written by a pre-M10 client, which had no concept of the
        // switch and whose values were being applied as if active.
        let bypass = d["bypass"] as? Bool ?? false

        lastAppliedSpeed = clampedSpeed
        lastAppliedBass = clampedBass
        lastAppliedReverb = clampedReverb
        lastAppliedPitch = clampedPitch
        lastAppliedBypass = bypass

        // Per-track memory (didSet → saveCurrentTrackSettings):
        //  - FOLLOWER: suppress (sync-audit-3.md F7). The values describe the
        //    OWNER's playback, not intent for whatever track is loaded here;
        //    persisting them corrupted that track's per-file memory.
        //  - OWNER: persist. A remote's slider drag is exactly a local drag of
        //    the song playing here and must stick the same way. Suppressing it
        //    meant the next track change restored the OLD memory and published
        //    it, silently undoing the other device's change — which read as
        //    "reverb/bass don't sync".
        player.isApplyingRemoteSettings = !ownsAudio()
        defer { player.isApplyingRemoteSettings = false }
        player.playbackSpeed = clampedSpeed
        player.bassBoost = clampedBass
        player.reverbAmount = clampedReverb
        player.pitchShift = clampedPitch
        player.effectsBypass = bypass
    }
}

File: desktop/src/settingsSync.ts
// Two-way effects-settings sync — twin of musicApp/Sync/SettingsSync.swift.
// Doc: users/{uid}/sync/settings (singleton, same doc family as session).
//   { speed, bassDb, reverbPct, pitchSt?, bypass?, updatedBy, at }
// LWW is Firestore's own snapshot ordering, same as the session/playlist
// docs — `at` (ServerClock ms) is carried for parity/debugging, not compared
// client-side. `updatedBy` filters same-device echo.
//
// What the doc means: "the effects currently AUDIBLE on the owning device".
// Any device may write it (a remote's slider drag is a live control of the
// owner); the owner also writes it whenever its audible values change for
// any other reason (per-track memory restore on a track change, reset).
import { Firestore, doc, setDoc, onSnapshot, Unsubscribe } from "firebase/firestore";
import { DEVICE_ID } from "./protocol";
import { serverClock } from "./serverClock";

export interface SettingsDoc {
  speed: number;     // playback rate multiplier
  bassDb: number;     // raw dB, same units both apps
  reverbPct: number;  // 0-100 — desktop's internal fx.reverb is a 0-1 fraction
  // Pitch shift in semitones, ±12, 0.5 st steps on both UIs. OPTIONAL on the
  // wire: absent means "written by a client from before pitch synced", read
  // as 0. Pitch was the one effect that never left the device it was set on
  // ("reverb, pitch and bass don't sync") — the iOS side simply never
  // published it and this side had it marked local-only.
  pitchSt?: number;
  // Effects master switch (iOS effectsBypass / desktop fx.bypass). Without it
  // a device with effects OFF still published its stored slider values, and a
  // device with effects ON applied them audibly — and they contradicted
  // PlaybackState.rate, which IS bypass-adjusted (sync-audit-4 M10).
  // OPTIONAL on the wire: absent means "an old client wrote this", and the
  // safe reading of a doc from a client that had no concept of bypass is
  // "not bypassed", which is exactly how those values were being applied.
  bypass?: boolean;
}

/** Wire ranges, shared with iOS (AudioSettingsSheet sliders). */
export const SETTINGS_RANGE = {
  speed: [0.5, 2.0], bassDb: [-10, 20], reverbPct: [0, 100], pitchSt: [-12, 12],
} as const;

const clamp = (v: number, [lo, hi]: readonly [number, number]) => Math.min(Math.max(v, lo), hi);

/** Parse + clamp a raw settings doc. Undefined when the doc is unusable
 *  (missing a required field, or this device's own echo). Pure — pinned by
 *  tests/settingsDoc.test.ts, and the single place the optional fields'
 *  defaults live. */
export function parseSettingsDoc(
  d: Record<string, unknown> | undefined, selfId: string = DEVICE_ID,
): Required<SettingsDoc> | undefined {
  if (!d || d.updatedBy === selfId) return undefined;
  if (typeof d.speed !== "number" || typeof d.bassDb !== "number"
      || typeof d.reverbPct !== "number") return undefined;
  return {
    speed: clamp(d.speed, SETTINGS_RANGE.speed),
    bassDb: clamp(d.bassDb, SETTINGS_RANGE.bassDb),
    reverbPct: clamp(d.reverbPct, SETTINGS_RANGE.reverbPct),
    pitchSt: typeof d.pitchSt === "number" ? clamp(d.pitchSt, SETTINGS_RANGE.pitchSt) : 0,
    bypass: typeof d.bypass === "boolean" ? d.bypass : false,
  };
}

export class SettingsSync {
  onRemote?: (s: Required<SettingsDoc>) => void;

  private unsub?: Unsubscribe;
  private uid = "";

  constructor(private db: Firestore) {}

  start(uid: string) {
    this.stop();
    this.uid = uid;
    this.unsub = onSnapshot(this.ref(), snap => {
      const s = parseSettingsDoc(snap.data());
      if (s) this.onRemote?.(s);
    });
  }

  stop() {
    this.unsub?.(); this.unsub = undefined;
    this.uid = "";
  }

  push(s: SettingsDoc) {
    if (!this.uid) return;
    void setDoc(this.ref(), { ...s, updatedBy: DEVICE_ID, at: serverClock.nowMs })
      .catch(() => {});
  }

  private ref() {
    return doc(this.db, "users", this.uid, "sync", "settings");
  }
}
